qi-test/tests/compiler/rules/full-cycle.rkt

```diff
@@ -46,7 +47,46 @@
                (deforested?
                  (phase1-eval
                   (qi-compile
-                   #'(~>> (filter odd?) values (map sqr)))))))))
+                   #'(~>> (filter odd?) values (map sqr)))))))
+   (test-suite
+    "sandboxed evaluation"
+    (test-not-exn "Plays well with sandboxed evaluation"
+                  ;; This test reproduces the bug and the fix fixes it. Yet,
+                  ;; coverage does not show the lambda in `my-emit-local-step`
+                  ;; as being covered. This could be because the constructed
+                  ;; sandbox evaluator "covering" the code doesn't count as
+                  ;; coverage by the main evaluator running the test?
+                  ;; We address this by putting `my-emit-local-step` in a
+                  ;; submodule, which, by default, are ignored by coverage.
+                  (lambda ()
+                    (let ([eval (parameterize ([sandbox-output 'string]
+                                               [sandbox-error-output 'string]
+                                               [sandbox-memory-limit #f])
+                                  (make-evaluator
+                                   'racket/base
+                                   '(require (for-syntax racket/base)
+                                             ;; necessary to recognize and expand core forms correctly
+                                             qi/flow/extended/expander
+                                             ;; necessary to correctly expand the right-threading form
+                                             qi/flow/extended/forms
+                                             syntax/macro-testing
+                                             racket/list
+                                             qi/flow/core/compiler
+                                             (submod qi/flow/extended/expander invoke))
```

Comment:
This feels like recreating a lot of the guts of Qi for the test—does `'(require qi)` not work? (The `(for-syntax racket/base)` might still be needed if you use a macro, but see later.)